crypto/s2n_certificate.c

```diff
@@ -176,30 +178,132 @@ struct s2n_cert_chain_and_key *s2n_cert_chain_and_key_new(void)
     GUARD_PTR(s2n_pkey_zero_init(chain_and_key->private_key));
     memset(&chain_and_key->ocsp_status, 0, sizeof(chain_and_key->ocsp_status));
     memset(&chain_and_key->sct_list, 0, sizeof(chain_and_key->sct_list));
-    chain_and_key->san_names = NULL;
-    chain_and_key->x509_cert = NULL;
+    chain_and_key->cn_names = s2n_array_new(sizeof(struct s2n_blob));
+    if (!chain_and_key->cn_names) {
+        return NULL;
+    }
+
+    chain_and_key->san_names = s2n_array_new(sizeof(struct s2n_blob));
+    if (!chain_and_key->san_names) {
+        return NULL;
+    }
 
     return chain_and_key;
 }
 
-static int s2n_cert_chain_and_key_set_x509(struct s2n_cert_chain_and_key *chain_and_key, struct s2n_blob *leaf_bytes)
+int s2n_cert_chain_and_key_load_sans(struct s2n_cert_chain_and_key *chain_and_key, X509 *x509_cert)
 {
-    const unsigned char *leaf_der = leaf_bytes->data;
-    X509 *cert = d2i_X509(NULL, &leaf_der, leaf_bytes->size);
-    if (!cert) {
-        S2N_ERROR(S2N_ERR_INVALID_PEM);
-    }
+    notnull_check(chain_and_key->san_names);
 
-    chain_and_key->x509_cert = cert;
-
-    GENERAL_NAMES *san_names = X509_get_ext_d2i(cert, NID_subject_alt_name, NULL, NULL);
+    GENERAL_NAMES *san_names = X509_get_ext_d2i(x509_cert, NID_subject_alt_name, NULL, NULL);
     if (san_names == NULL) {
         /* No SAN extension */
         return 0;
     }
 
-    chain_and_key->san_names = san_names;
+    for (int i = 0; i < sk_GENERAL_NAME_num(san_names); i++) {
+        GENERAL_NAME *san_name = sk_GENERAL_NAME_value(san_names, i);
+        if (!san_name) {
+            continue;
+        }
 
+        if (san_name->type == GEN_DNS) {
+            /* Per https://www.openssl.org/docs/man1.1.0/man3/ASN1_STRING_data.html there may
+             * be embedded NULLs inside of the SAN string. The comparison later with any proper
+             * DNS hostnames will safely fail.
+             */
+            unsigned char *san_str = san_name->d.dNSName->data;
```

Comment:
I'm not following this comment? Where does ASN1_STRING_data come into the picture? Are we supposed to call it on san_name->d.dNSName?

Is it possible to create a test case with an embedded NULL in the SAN to corroborate the statement that it will safely fail?

Reply:
Poked around the openssl sources again and we can safely assume it's `V_ASN1_IA5STRING` so this comment doesn't apply. openssl's host matching function doesn't worry about embedded NULLs for SAN names: https://github.com/openssl/openssl/blob/4286ca47c7fac2b5a6208b377967e0f095e9abc5/crypto/x509v3/v3_utl.c#L793-L798

Reply:
I'll update the comment in code

Reply:
Updated comment to mention this is `V_ASN1_IA5STRING`